### ISSUE TITLE
Iterate over List in Blazor

### ISSUE THREAD

**Issue by Sm1le291** [dotnet/aspnetcore#18904]
Hi everyone! I faced with strange behavior of Blazor.
For example we have component as below:
`@page "/"
@using BlazingPizza.ViewModels

@inject HttpClient HttpClient 

<h1>Blazing Pizzas</h1>

@if (_specials != null)
{
    @foreach (var pizza in _specials)
    {
        @pizza.Description
    }
}

@code {

    List<PizzaSpecial> _specials; // = Enumerable.Empty<ViewModels.PizzaSpecial>();

    protected async override Task OnInitializedAsync()
    {
        Console.WriteLine($"Time from handler: {DateTime.Now.ToString()}:{DateTime.Now.Millisecond.ToString()}");
        var result = await HttpClient.GetJsonAsync<List<ViewModels.PizzaSpecial>>("specials");

        _specials = result.Select(x => new PizzaSpecial()
        {
            Id = x.Id,
            Description =  x.Description
        }).ToList();
        
        
    }

}`

if we remove "if (_specials != null)" check we show an exceprtion NullReferenceException.
But code in OnInitializedAsync executes earlier than iteration over collection.
Could you explain why does it happen?

**Comment by rynowak** [comment]
This is discussed here in the documentation: https://docs.microsoft.com/en-us/aspnet/core/blazor/lifecycle?view=aspnetcore-3.1#handle-incomplete-async-actions-at-render

Let us know if this isn't a clear enough explanation.

TLDR: Components need to be able to render without waiting for things like HTTP requests to complete. You need to think about what you want to be displayed while the specials are loading.

**Comment by Sm1le291** [comment]
@rynowak thanks for answer! It partly described, what happened, but I still do not understand one thing. Why at the second time when component re-renders we for example fall into cycle and iterate  through the items of collection which was null for the first time, but still show an exception message.
Does it mean that if exception happened while first render, code on re render can executes but can not update the screen.
That's exactly what happens in my case, I show messages from cycle in console but continue to show error on   main browser screen

**Comment by rynowak** [comment]
>Does it mean that if exception happened while first render, code on re render can executes but can not update the screen. 

Here's what I suspect is happening.

- `OnInitializedAsync` starts running. It runs up to the yield point `await HttpClient.GetJsonAsync<>()`
- When that await occurs control is yielded to the renderer, which will then render your component
- Rendering the component hits a null reference exception
- This updates the UI to show the error bar - from this point on the app is "broken" and nothing will update the UI.
- After the HTTP request completes `OnInitializedAsync` continues from `await HttpClient.GetJsonAsync<>()` and runs to completion
- No further rendering occurs because the app has already crashed

So basically, encountering an unhandled exception will cause the UI to crash and prevent further updates, but will not abort any tasks that are currently in flight (`OnInitializedAsync`).

From my point of view there's nothing unexpected happening here.

**Comment by Sm1le291** [comment]
@rynowak thanks for explanation. Everything clear!
